### PR TITLE
Implement store info reporting and recover plan execution logic for o…

### LINKED ISSUE #10483: Support online recovery for Raft majority failure
## Development Task

Currently, we have a tool to recover Raft groups from majority failure:
```
tikv-ctl unsafe-recover remove-fail-stores -s <store_id, store_id> …
tikv-ctl recreate-region -r <region_id> ...
```

The first command is used to recover Raft groups of which the majority but not all replicas failed. Its mechanism is removing failed stores from membership configurations of those Raft groups.

The second command is used to recover Raft groups of which all replicas failed. Before running the command we need to find those Raft groups, which is not a documented routine. So there are no users who can do this recovery without our DBAs’ help.
Both of the 2 commands should be run on stopped TiKV instances, which makes the recovery process so long and complex that not all users can perform all steps correctly.

And there are some corner cases that could lead to inconsistent membership configuration among all replicas. If the Region splits later, this mistake will lead to inconsistent Region boundaries, which is very hard to locate and fix.
So we need to re-design and re-implement a new feature to recover a TiKV cluster from majority failure. The feature should be:
- able to be executed online
- able to avoid all inconsistencies
- a feature instead of a tool, focus on Raft majority failure, shouldn’t be used for other situations
- easy to use, at most 1 or 2 steps or command

### Specs
We must know there are no silver bullets for Raft majority failure. There must be something lost after a TiKV cluster recovers from that. Considering such integrations or consistencies:
- Committed writes  can be lost, so linear consistency of Raft can be broken;
- Some prewrites can be lost for a given transaction, so Transaction Atomicity and/or Consistency can be broken;
- Commits can be lost for a given transaction, so Transaction Consistency and/or Durability can be broken;
- For a given table, both records or indices can be lost, so Data Consistency can be broken;
- For a given transaction, it can be partially committed, so Data Integrity can be broken.

#### Stage 1: Implement basic online recovery 
- [x] Add `recover-store-failure` command of pd-ctl https://github.com/v01dstar/pd/pull/1
- [x] Add framework of unsafe recovery worker in PD  https://github.com/v01dstar/pd/pull/2
- [x] Support `DumpUnavailableRegions` command
  - [x] Change `StoreHeartbeatRequest` and `StoreHeartbeatResponse` in kvproto  https://github.com/v01dstar/kvproto/pull/1
  - [x] Dispatch it by `StoreHeartbeatResponse` in PD https://github.com/v01dstar/pd/pull/2
  - [x] Handle it and return results in TiKV https://github.com/v01dstar/tikv/pull/1
- [x] Implement the logic of unsafe recovery worker in PD(only keep one replica) https://github.com/tikv/pd/pull/4238
- [x] Support `UnsafeRecovery` command https://github.com/tikv/tikv/pull/11146
  - [x] Add a new API in kvproto
  - [x] Implement logic in TiKV
- [x] Support `RecreateRegion` command https://github.com/tikv/tikv/pull/11146
  - [x] Add a new API in kvproto
  - [x] Implement logic in TiKV 
- [x] Test [5d]
- [x] Add recovery document https://github.com/pingcap/docs-cn/pull/7381

#### Stage 2: Make the best effort for the case of replicas more than 3
- [x] Forbid the removed stores starting again https://github.com/tikv/pd/pull/4338 @v01dstar 
- [x] PD disables the schedule operations in the recovery process https://github.com/tikv/pd/pull/4475 @v01dstar 
- [x] TiKV makes sure commit index == apply index https://github.com/tikv/tikv/pull/11716 @v01dstar ‘
  - [x] Fix snapshot corner case https://github.com/tikv/tikv/pull/12308 @v01dstar 
- [x] Avoid PD issuing peer report again and again https://github.com/tikv/pd/pull/4705 @v01dstar 
- [x] Dispatch recovery plan by raft and execute in apply https://github.com/tikv/tikv/pull/12022 @v01dstar
  - [x] Modify proto https://github.com/pingcap/kvproto/pull/904 @Connor1996 
  - [x] Report unapplied commit merge, to let force leader on region of commit merge first https://github.com/tikv/tikv/pull/12638
  - [x] Fix https://github.com/tikv/tikv/issues/12506
- [x] Adjust the logic of unsafe recovery worker in PD and divide the process into multiple stages https://github.com/tikv/pd/pull/4704 @Connor1996 
  - [ ] Make region tree generic (optional)
  - [ ] Send store heartbeat response actively (optional)
  - [x] Tombstone tiflash learner peer if it's latest https://github.com/tikv/pd/pull/5045
  - [x] Fix blocking on exit force leader https://github.com/tikv/pd/pull/4982
- [x] Implement `force_leader` https://github.com/tikv/tikv/pull/11932 @Connor1996 
  - [x] Fix hibernated peer case https://github.com/tikv/tikv/pull/12364
  - [x] Fix related flaky tests and add back exit force leader check https://github.com/tikv/tikv/pull/12810
  - [ ] Fix corner case: supposing (a, b, c, d, e) -> (a, b, c, d, f) -> (a, b, c, g, f), and a, b, g, f all fail. c is isolated during the whole time. It may think it still has quorum (c, d, e)
  - [x] Add `RecoveryInProgress` error https://github.com/pingcap/kvproto/pull/863 @Connor1996
  - [x] Handle `RecoveryInProgress` error in client https://github.com/tikv/client-go/pull/485 @Connor1996
  - [x] Dynamic change `check_quorum` of raft https://github.com/tikv/raft-rs/pull/461 @v01dstar 
- [ ] ~~Invalidate region cache in clients after recovery (optinal)~~
- [x] Public RFC https://github.com/tikv/rfcs/pull/91
- [x] Update doc https://github.com/pingcap/docs-cn/pull/9495
- [x] Adaption of Tiflash https://github.com/pingcap/tidb-engine-ext/pull/62
- [x] Test

#### Stage 3: Enhance inconsistency fix tools (Optional）
- [ ] Adjust `recover-mvcc` tool to handle the case the primary key lock and secondary key locks are not consistent after unsafe recovery
- [ ] Make TiDB admin recover table differential
  - [ ] Need more detailed design,TBD

### PR DESCRIPTION
**Pull request by v01dstar**
### What problem does this PR solve?

Issue Number: close #10483  

Problem Summary:
Support online recovery for Raft majority failure

### What is changed and how it works?

Proposal: [Design Doc](https://docs.google.com/document/d/1oJl4vgtvuIKBQXtlwhpmt3duI0D1Fm5HbhwdbEjkKm0/edit#heading=h.hvo12ov0sqrj) 

What's Changed:
1. Several unsafe recover related PD commands are added.  They are: `unsafe remove-failed-store <store id>,<store id>...`, `unsafe remove-failed-store show` and `unsafe remove-failed-store history`
2. After PD received unsafe recovery request, it notifies all stores through StoreHeartbeatResponse to send a detailed report (local region and raft info) in the next heartbeat.
3. Once PD have all the reports, it generates a recover plan and sends them to each individual store (using StoreHeartbeatResponse as well) to execute.
4. The recover is considered finished if all the target states are met (by comparing following store reports with recovery plan).

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`: WIP
- PR to update `pingcap/tidb-ansible`: WIP
- Need to cherry-pick to the release branch: **Not applicable**

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression:  **Not applicable**
    - Consumes more CPU:  **Not applicable**
    - Consumes more MEM: **Not applicable**
- Breaking backward compatibility: **Not applicable**

### Release note 

```release-note
Support online recovery for Raft majority failure
```

### CODE REVIEW

File: tests/integrations/raftstore/test_unsafe_recover.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    cluster.run();
+    let nodes = Vec::from_iter(cluster.get_node_ids());
+    assert_eq!(nodes.len(), 3);
+
+    let pd_client = Arc::clone(&cluster.pd_client);
+    // Disable default max peer number check.
+    pd_client.disable_default_operator();
+
+    let election_timeout = configure_for_lease_read(&mut cluster, None, None);
+    cluster.add_send_filter(IsolationFilterFactory::new(nodes[1]));
```
Comment (Connor1996) [inline tests/integrations/raftstore/test_unsafe_recover.rs:24]: can we just call `stop_node`

**Review by Connor1996 — APPROVED** [review]
LGTM

**Review by hicqu — APPROVED** [review]

**Pull request merged**